### PR TITLE
Tabbed function

### PR DESCRIPTION
**Pull request by huijunchen9260**
I modified the X11 window and added tabbed support.

### CODE REVIEW

File: autoload/slime.vim [hunk 1]
```diff
@@ -229,8 +229,9 @@ endfunction
 function! s:X11Config() abort
   if !exists("b:slime_config")
     let b:slime_config = {"window_id": ""}
-  end
+  endif
```
Comment (jpalardy) [inline autoload/slime.vim:232]: whoah… did that never work?

File: autoload/slime.vim [hunk 2]
```diff
@@ -248,6 +249,22 @@ function! s:DtachConfig() abort
...
+
+function! s:TabbedSend(config, text)
+  call system("xdotool type --delay 0 --window " . b:slime_config["window_id"] . " -- " . shellescape(a:text))
+endfunction
+
+function! s:TabbedConfig() abort
+  if !exists("b:slime_config")
+    let b:slime_config = {"window_id": ""}
+  endif
+  let b:slime_config["window_id"] = trim(system("xwininfo -children -id $XEMBED | grep '     0x' | sed -e's@^ *\\(0x[0-9a-f]*\\) \"\\([^\"]*\\)\".*@\\1 \\2@' | tail -n +2 | dmenu -i -l 5 -p 'REPL window: ' | cut -b 1-9 | xargs printf '%d'"))
```
Comment (jpalardy) [inline autoload/slime.vim:264]: this … is quite a line 😐 

I don't think I can merge/support this. Let me add a comment to discuss alternatives.

File: autoload/slime.vim (hunk 1 above)
Comment (huijunchen9260) [reply autoload/slime.vim:232]: Surprising work LOL

**Comment by jpalardy** [comment]
hi @huijunchen9260,

While I do not doubt that your PR works, I'm hesitant to add more complexity to vim-slime itself.

All the plugins have a few things in common:
- they prompt user for config
- they (usually) feed the config through some command-line invocation
- they trust that text goes where it's supposed to go

To the extent that "a command-line invocation" is a black box and could be anything, maybe that's a property worth exploring.

What if:
- there as a new "generic" (or "custom") plugin
- it could be configured to ask custom prompts for config
- and turn around to feed it to a custom (configured) command-line script

In practice, you could take

```bash
xwininfo -children -id $XEMBED | grep '     0x' | sed -e's@^ *\\(0x[0-9a-f]*\\) \"\\([^\"]*\\)\".*@\\1 \\2@' | tail -n +2 | dmenu -i -l 5 -p 'REPL window: ' | cut -b 1-9 | xargs printf '%d'"
```

and put it in a script, maybe named `send-to-x11-tabbed`, and the configuration could look like

```vim
let g:slime_target = "custom"
let g:slime_custom_config_keys = ["window_id"]
let g:slime_custom_script = "send-to-x11-tabbed"
```
all that's missing is a way to feed `window_id` to the custom_script 🤔

File: autoload/slime.vim (hunk 2 above)
Comment (huijunchen9260) [reply autoload/slime.vim:264]: Sure. I don't know vim script, so what I can do is just shell scripting.

**Comment by huijunchen9260** [comment]
> ```viml
> let g:slime_target = "custom"
> let g:slime_custom_config_keys = ["window_id"]
> let g:slime_custom_script = "send-to-x11-tabbed"
> ```
> 
> all that's missing is a way to feed `window_id` to the custom_script

The way my tabbed implementation on getting windows id is by the long script above. The way the X11 window implementation to get windows id is by `xdotool selectwindow`. I think such implement do not need to specify window id, but need to specify the script?

File: autoload/slime.vim [hunk 3]
```diff
@@ -250,21 +250,18 @@ function! s:DtachConfig() abort
...
-function! s:TabbedConfig() abort
+function! s:CustomConfig() abort
   if !exists("b:slime_config")
     let b:slime_config = {"window_id": ""}
   endif
-  if !exists("b:slime_custom_script")
-    let b:slime_custom_script = tabbed
-  endif
-  let b:slime_config["window_id"] = trim(system("b:slime_custom_script"))
+  let b:slime_config["window_id"] = trim(system("tabbed-slime"))
```
Comment (huijunchen9260) [inline autoload/slime.vim:264]: I am not sure how to add a global variable in vim, and utilize it in `trim(system())`. But I export those long code into the `tabbed-slime` script, and do some modification. The content of `tabbed-slime` is:

```
xwininfo -children -id $XEMBED | grep '     0x' | awk '{print $1, $2}' | tail -n +2 | dmenu -i -l 5 -p 'REPL window: ' | cut -b 1-9 | xargs printf '%d'
```

Could I ask how to assign a global variable and put it in this chunk of vim script?

Thank you!

Comment (jpalardy) [reply autoload/slime.vim:264]: `let g:var_name = 5` sets a global var

same for using it: `let b:slime_config["window_id"] = trim(system(g:var_name))`

**Comment by huijunchen9260** [comment]
OK. With your advice, I fully integrate the custom script into the X11 window module and allow X11 function to execute external script to obtain windows id.

File: autoload/slime.vim [hunk 4]
```diff
@@ -229,8 +229,14 @@ endfunction
...
-  end
-  let b:slime_config["window_id"] = trim(system("xdotool selectwindow"))
+  endif
+  if exists("g:slime_custom_script")
+    let b:slime_action = g:slime_custom_script
+  else
+    let b:slime_action = "xdotool selectwindow"
+  endif
+  let b:slime_config["window_id"] = trim(system(b:slime_action))
+  echom b:slime_config["window_id"]
```
Comment (jpalardy) [inline autoload/slime.vim:239]: let's remove that debug `echom`

**Comment by jpalardy** [comment]
hi @huijunchen9260,

It's looking good 😄 

Let me do a pass over the weekend to see if I'm going to generalize this even more. (I wish this weren't vimscript… I have a lot of ideas on how to break things into "modules", but I'm hitting the vimscript wall)

**Comment by huijunchen9260** [comment]
I am just avoiding vim scripting by shell scripting because I don't know much about it LOL
Hope you can come up a good way to generalize this fantastic plugin!

**Comment by jpalardy** [comment]
closing because of #237

**Comment by jpalardy** [comment]
updated the README

if you could read [the new section](https://github.com/jpalardy/vim-slime#advanced-configuration-overrides) and let me know what you think, I would appreciate it.

**Comment by huijunchen9260** [comment]
I wonder...is it possible to sort of "debug"? Like set the breakpoint to see what happened if there is such a debugger for this language.

Now I am learning matlab, and from [this github](https://github.com/MortenStabenau/matlab-vim/blob/master/autoload/matlab.vim), I realize that I can probably just add the breakpoint function with just vimm-slime. Here's my thought (I still know nothing about vim-script lol)

At [L68](https://github.com/MortenStabenau/matlab-vim/blob/master/autoload/matlab.vim#L68) it shows that adding a breakpoint is like

```matlab
dbclear <<filename>>; dbstop <<filename>> at <<line number>>
```
and also some action to clean the debug.

Is it possible to set a keybinding to generate these matlab debugging code, and send it to the REPL using vim-slime?

**Comment by jpalardy** [comment]
Hi @huijunchen9260 

If what you're trying to do is send `dbclear … etc … at line number X`, I don't see why that wouldn't be possible.

You can check [the bindings](https://github.com/jpalardy/dotfiles/blob/master/vim/plugins.vim#L24-L26) from my dotfiles for examples on how to send specific strings through vim-slime.

You'll probably have to write a function, that calls `SlimeSend0`, and map a key binding to call that function.

Let me know how that goes.

**Comment by huijunchen9260** [comment]
Your vimrc is so awesome that forced me to rewrite my ugly and messy vimrc lol
and...I succeed implementing that debugging setting with my tabbed function using vim-slime.

vim-slime is such a general purpose and useful tool! Thank you again for creating this!!!

For someone who is interested in my setup, this is what I do for future refernce:

```vimL
"Matlab
" Send breakpoint to REPL
let f = substitute(expand('%:t'), '\.m', '', 'g')
" Run the file
autocmd Filetype matlab nmap <c-s>r :SlimeSend0 'run("'.f.'");' <CR>
" Add a breakpoint at current line
autocmd Filetype matlab nmap <c-s>a :SlimeSend0 'dbstop '.f.' at '.line('.').';' <CR>
" Clear all breakpoints
autocmd Filetype matlab nmap <c-s>b :SlimeSend0 'dbclear '.f.';' <CR>
```

**Comment by jpalardy** [comment]
happy that worked out 😄

**Comment by huijunchen9260** [comment]
Btw, I want to ask why I have to map `SlimeSend0`  as `nmap` but not `nnoremap`? I don't know why but it just says error.

Especially `nmap <c-s>c <Plug>SlimeConfig`. Somehow `nnoremap` does work.

My corresponding part of vimrc:

```vimL
xmap <c-s><c-s> <Plug>SlimeRegionSend
nmap <c-s><c-s> <Plug>SlimeParagraphSend
nmap <c-s>l <Plug>SlimeLineSend
nmap <c-s>c <Plug>SlimeConfig
" X11 -> tabbed
let g:slime_target = "x11"
function! SlimeOverrideConfig()
    let b:slime_config = {"window_id": ""}
    let b:slime_config["window_id"] = system("tabbed-slime")
endfunction
```

**Pull request closed without merging**